* Added EKS config + validation, kubeconfig output

JadeHayes opened the pull request:
## Tasks

- [x] Create a dedicated AWS VPC
- [x] Create an EKS cluster inside the VPC
- [x] Create an ECR repository
- [x] Create an IAM policy that allows access to your ECR repository and add it to the IAM role assigned to the EKS nodes
- [x] Expose inputs for **region** and **kubernetes_version**
- [x] Export outputs for **kubeconfig** and **repository**
- [x] Verify that **kubeconfig** allows admin access to EKS cluster
- [x] Verify that you can access your ECR repo from inside a container running in the cluster

* Generated EKS resource creation.

* terraform plan is running successfully.

* TODO: need to add IAM roles + permissions.

Commit:
* Added EKS config + validation and mapping to HCL.

* Generated EKS resource creation.

* terraform plan is running successfully.

* TODO: need to add IAM roles + permissions.

sonnysideup commented on resources/terraform/aws-eks/outputs.tf:
```diff
@@ -0,0 +1,5 @@
+  description = "Kubeconfig for eks cluster"
+  sensitive = true
+  value       = local.kubeconfig
+}
```
format and newline

sonnysideup commented on resources/terraform/aws-eks/main.tf:
```diff
@@ -0,0 +1,169 @@
+    protocol  = "-1"
+    self      = "true"
+  }
+}
```
add newline at the end

sonnysideup commented on resources/terraform/aws-eks/main.tf:
```diff
@@ -0,0 +1,169 @@
+      version = "~> 4.0"
+    }
+  }
+}
```
I think this can be omitted for now because the `"terraform-aws-modules/eks/aws"` imposes a constraint on this provider. we may need it later, but let's cross that road if we get there.

sonnysideup commented on resources/terraform/aws-eks/main.tf:
```diff
@@ -0,0 +1,169 @@
+
+locals {
+  name        = "testenv-eks-${random_string.name_suffix.result}"
+  vpc_id = aws_vpc.this.id
```
You can remove this local var and use `aws_vpc.this.id` directly

sonnysideup commented on resources/terraform/aws-eks/main.tf:
```diff
@@ -0,0 +1,169 @@
+  version = "~> 18.0"
+
+  cluster_name    = "${local.name}-cluster"
+  cluster_version = "1.22"
```
```suggestion
  cluster_version = var.kubernetes_version
```

Commit:
* Using VPC module instead of vpc resource.

* Reformatted tf files.

Commit:
* Updated kubernetes version.

Commit:
* Added ecr module for ecr repo.

* Updated terraform output to include ecr repo url.

Commit:
* Updated IAM policy + role.

Commit:
* Added kubeconfig path output variable.

* Added kubeconfig as a null resource to update the kubeconfig.

* Added nodes to the EKS cluster.

Commit:
* Added IAM for node access.

Commit:
* Added egress.

* Changed manage_aws_auth_configmap back to true.

* Enabled hostnames though I am unsure if I need that.

Commit:
* Added AWS & Kubernetes providers.

Commit:
* Added ecr policy to node groups.

JadeHayes committed:
Update outputs.tf

- Updated description to specify it's the kubeconfig path and not the actual kubeconfig.

JadeHayes committed:
Update variables.tf

- Updated kubernetes version description.

sonnysideup commented on resources/terraform/aws-eks/main.tf:
```diff
@@ -0,0 +1,185 @@
+    api_version = "client.authentication.k8s.io/v1"
+    command     = "aws"
+    # This requires the awscli to be installed locally where Terraform is executed
+    args = ["eks", "get-token", "--cluster-name", module.eks.cluster_id]
```
You'll might need to leverage the `--region` flag with this command since there's no guarantee the CI machines will have that set and I'm unsure if the `aws` command will inherit the settings of `provider "aws"`, but I highly doubt it.

sonnysideup commented on resources/terraform/aws-eks/outputs.tf:
```diff
@@ -0,0 +1,10 @@
+  value       = var.kubeconfig_path
+}
+
+output "repo_url" {
```
```suggestion
output "repository" {
```

It might make sense for all of these repository output names to match (see https://github.com/dominodatalab/testenv/blob/3275dc58c941410a021285b0acba681cd8350ef4/resources/terraform/gcp-gke/outputs.tf#L7) for consistency's sake. It's not that we can't map in the Go types, but it just seems like since we control the output name, then why not make them uniform, right?

sonnysideup commented on resources/terraform/aws-eks/variables.tf:
```diff
@@ -0,0 +1,17 @@
+variable "kubernetes_version" {
+  type        = string
+  description = "The Kubernetes version."
+  default     = "1.22"
```
```suggestion
  default     = "1.23"
```

Bump to the [latest supported version](https://docs.aws.amazon.com/eks/latest/userguide/kubernetes-versions.html)? It's not a huge deal but, as AWS removes support for older versions, eventually `1.22` will no longer be valid. The same is true for `1.23` but at least that buys us more runway before we have to update this value.

sonnysideup commented on resources/terraform/aws-eks/variables.tf:
```diff
@@ -0,0 +1,17 @@
+}
+
+
+variable "kubeconfig_path" {
+  type        = string
+  description = "Path to the kubeconfig file."
+  default     = "./kubeconfig"
+}
```
I don't think we want this variable here since we always want to generate the `kubeconfig` inside the module root. I would remove this and use `path.module` to ensure it gets created where it should be. See https://github.com/dominodatalab/testenv/blob/3275dc58c941410a021285b0acba681cd8350ef4/resources/terraform/gcp-gke/main.tf#L71-L75 for the details.

This is also means you'll need to change the output. See https://github.com/dominodatalab/testenv/blob/3275dc58c941410a021285b0acba681cd8350ef4/resources/terraform/gcp-gke/outputs.tf#L1-L5

sonnysideup commented on resources/terraform/aws-eks/main.tf:
```diff
@@ -0,0 +1,185 @@
+  name = module.eks.cluster_id
+}
+
+module "vpc" {
+  source             = "terraform-aws-modules/vpc/aws"
+  version            = "~> 3.18"
+  enable_nat_gateway = true
+  cidr               = "10.0.0.0/16"
```
This is not a big deal but worth pointing since we're trying to make these TF files consistent.

If you see your `module "eks"`, the `source` and `version` are always grouped together at the top and are separate from the other modules variables. I like this style and use it personally since it creates a visually separation between (a) what is the module/version and (b) what are the inputs.

I would suggest making all your modules follow this format.

```terraform
module "eks" {
  source  = "terraform-aws-modules/eks/aws"
  version = "~> 18.30"

  ...other variables
```

sonnysideup reviewed (changes requested):
Looks good @JadeHayes, but it needs a couple of tweaks.

1. Make the tweaks
2. Make sure to **Resolve** the conversations

Commit:
* Removed kubeconfig path from output in favor of local variable.

* Updated k8s version to 1.23

* Updated module format to match other tf modules in our repo.

Commit:
Merge branch 'jade/add-eks-test-env' of github.com:dominodatalab/testenv into jade/add-eks-test-env
